File: server/auth/serviceKey.ts
import crypto from 'crypto'

export class ServiceKeyError extends Error {
  constructor(message: string) {
    super(message)
    this.name = 'ServiceKeyError'
  }
}

export function verifyServiceKey(header: string | undefined, secret: string): void {
  if (!secret) {
    throw new ServiceKeyError('Service key not configured')
  }
  if (!header) {
    throw new ServiceKeyError('Service key required')
  }

  const h = Buffer.from(header)
  const s = Buffer.from(secret)

  if (h.length !== s.length) {
    throw new ServiceKeyError('Invalid service key')
  }

  if (!crypto.timingSafeEqual(h, s)) {
    throw new ServiceKeyError('Invalid service key')
  }
}

File: python_server/auth/service_key.py
import hmac


class ServiceKeyError(Exception):
    """Raised when a service key is missing or invalid."""


def verify_service_key(header: str, secret: str) -> None:
    """Constant-time compare the X-Service-Key header against the configured secret."""
    if not secret:
        raise ServiceKeyError("Service key not configured")
    if not header:
        raise ServiceKeyError("Service key required")
    if not hmac.compare_digest(header, secret):
        raise ServiceKeyError("Invalid service key")
